### PR TITLE
Rewrite internal logic of block sync client

### PR DESCRIPTION
**Pull request by schomatis**
While trying not to modify too much of the external callers code. The only major change outside of `chain/blocksync/` is extracing the `FetchMessagesByCids`/`FetchSignedMessagesByCids` functions to the consumer. They did not actually used this protocol, just the `BlockService` embedded in one of its structures (that was also extracted since the server only needs the `ChainStore`).

### CODE REVIEW

**Comment by schomatis** [comment]
Not sure why `TestAPIRPC/testMining` is failing.

**Comment by magik6k** [comment]
> TestAPIRPC/testMining

Try rebasing / merging newer `next` - https://github.com/filecoin-project/lotus/issues/2740

**Comment by schomatis** [comment]
@magik6k same error

**Comment by Kubuxu** [comment]
Working on review of this.

File: chain/blocksync/client.go [hunk 1]
```diff
@@ -0,0 +1,445 @@
...
+					i, i+1)
+				// FIXME: Maybe give more information here, like CIDs.
+			}
+		}
+	}
+
+	if options.IncludeMessages {
+		validRes.messages = make([]*CompactedMessages, resLength)
+		for i := 0; i < resLength; i++ {
+			if res.Chain[i].Messages == nil {
```
Comment (whyrusleeping) [inline chain/blocksync/client.go:206]: what if there actually arent any messages?

Comment (schomatis) [reply chain/blocksync/client.go:206]: you'd have an empty structure but not a nil pointer

Comment (schomatis) [reply chain/blocksync/client.go:206]: but that's a good question, i now realize i had the implicit (unfounded) assumption that blocks couldn't be empty

File: chain/blocksync/client.go [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+	if err != nil {
+		client.RemovePeer(peer)
+		return nil, xerrors.Errorf("failed to open stream to peer: %w", err)
+	}
+
+	// Write request.
+	_ = stream.SetWriteDeadline(time.Now().Add(WRITE_REQ_DEADLINE))
+	if err := cborutil.WriteCborRPC(stream, req); err != nil {
+		_ = stream.SetWriteDeadline(time.Time{})
+		// FIXME: What's the point of setting a blank deadline that won't time out?
```
Comment (whyrusleeping) [inline chain/blocksync/client.go:381]: Yes, this is the way to clear a deadline

File: chain/blocksync/client.go [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+		//  Is this our way of clearing the old one?
+		client.peerTracker.logFailure(peer, build.Clock.Since(connectionStart))
+		// FIXME: Should we also remove peer here?
```
Comment (whyrusleeping) [inline chain/blocksync/client.go:384]: i currently assume the peer tracker won't try the peer again unless they are our only other option

File: chain/blocksync/client.go [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+		return nil, err
+	}
+	// FIXME: Same, why are we doing this again here?
+	_ = stream.SetWriteDeadline(time.Time{})
```
Comment (whyrusleeping) [inline chain/blocksync/client.go:388]: the previous call is inside of an error handler, we need to reset the deadline in either case (error or no error)

File: chain/blocksync/protocol.go [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+const READ_RES_DEADLINE = WRITE_REQ_DEADLINE
+const READ_RES_MIN_SPEED = 50<<10
+const SHUFFLE_PEERS_PREFIX = 5
+const WRITE_RES_DEADLINE = 60 * time.Second
+
+// FIXME: Rename. Make private.
+type Request struct {
+	// List of ordered CIDs comprising a `TipSetKey` from where to start
+	// fetching backwards.
+	// FIXME: Why don't we send a `TipSetKey` instead of converting back
```
Comment (whyrusleeping) [inline chain/blocksync/protocol.go:40]: tipsetkey didnt exist when we wrote this, mostly

File: chain/blocksync/protocol.go [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+// of grouped by block to save space, since there are normally
+// many repeated messages per tipset in different blocks.
+//
+// `BlsIncludes`/`SecpkIncludes` matches `Bls`/`Secpk` messages
+// to blocks in the tipsets with the format:
+// `BlsIncludes[BI][MI]`
+//  * BI: block index in the tipset.
+//  * MI: message index in `Bls` list
+//
+// FIXME: The logic to decompress this structure should belong
```
Comment (whyrusleeping) [inline chain/blocksync/protocol.go:144]: 👍

File: chain/blocksync/server.go [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+	log.Infow("block sync request",
+		"start", req.Head, "len", req.Length)
+
+	resp, err := server.processRequest(ctx, &req)
+	if err != nil {
+		log.Warn("failed to process request: ", err)
+		return
+	}
+
+	_ = stream.SetDeadline(time.Now().Add(WRITE_RES_DEADLINE))
```
Comment (whyrusleeping) [inline chain/blocksync/server.go:72]: need to clear the deadline too

Comment (schomatis) [reply chain/blocksync/server.go:71]: addressed in https://github.com/filecoin-project/lotus/pull/2715/commits/65512193ccf97e6fd9b22c9953086a1ea3ad967a

File: chain/blocksync/client.go [hunk 8]
```diff
@@ -378,14 +378,12 @@ func (client *BlockSync) sendRequestToPeer(
...
-		// FIXME: What's the point of setting a blank deadline that won't time out?
-		//  Is this our way of clearing the old one?
 		client.peerTracker.logFailure(peer, build.Clock.Since(connectionStart))
 		// FIXME: Should we also remove peer here?
 		return nil, err
 	}
-	// FIXME: Same, why are we doing this again here?
-	_ = stream.SetWriteDeadline(time.Time{})
+	_ = stream.SetWriteDeadline(time.Time{}) // clear deadline // FIXME: Needs
+	//  its own API (https://github.com/libp2p/go-libp2p-core/issues/162).
```
Comment (Kubuxu) [inline chain/blocksync/client.go:386]: Clearing Deadline by empty time is the Go's standard, I agree specific API might be nice.

File: chain/blocksync/client.go [hunk 9]
```diff
@@ -0,0 +1,443 @@
...
+			return nil, xerrors.Errorf("no peers available")
+		}
+	}
+
+	// Try the request for each peer in the list,
+	// return on the first successful response.
+	// FIXME: Doing this serially isn't great, but fetching in parallel
+	//  may not be a good idea either. Think about this more.
+	startTime := build.Clock.Now()
+	// FIXME: Should we track time per peer instead of a global one?
```
Comment (Kubuxu) [inline chain/blocksync/client.go:100]: Global times to used to track "what is the expected time we will need to get a response if this client fails us"

File: chain/blocksync/client.go [hunk 10]
```diff
@@ -0,0 +1,443 @@
...
+	//  Maybe we need to declare `res` in the signature.
+	if span.IsRecordingEvents() {
+		span.AddAttributes(
+			trace.Int64Attribute("resp_status", int64(res.Status)),
+			trace.StringAttribute("msg", res.ErrorMessage),
+			trace.Int64Attribute("chain_len", int64(len(res.Chain))),
+		)
+	}
+
+	client.peerTracker.logSuccess(peer, build.Clock.Since(connectionStart))
```
Comment (Kubuxu) [inline chain/blocksync/client.go:408]: We should really log a success only after we validate the response. It might be a bit hard to do.

**Comment by schomatis** [comment]
Still a lot of tests failing from unrelated reasons..

File: chain/blocksync/protocol.go (hunk 6 above)
Comment (jaceppo) [reply chain/blocksync/protocol.go:145]: jacepop

**Pull request merged**